R/renv.R

```diff
@@ -0,0 +1,86 @@
+translateRenvToPackrat <- function(bundleDir) {
+  renv <- jsonlite::read_json(renvLockFile(bundleDir))
+  packrat <- packratFromRenv(renv)
+
+  packratPath <- packratLockFile(bundleDir)
+  dir.create(dirname(packratPath), showWarnings = FALSE)
+  write.dcf(packrat, packratPath)
+}
+
+packratFromRenv <- function(renv) {
+  meta <- packratMeta(renv)
+  packages <- packratPackages(renv$Packages)
+  rbind_fill(list(meta, packages))
+}
+
+packratMeta <- function(renv) {
+  repos <- vapply(
+    renv$R$Repositories,
+    function(x) paste0(x$Name, "=", x$URL),
+    character(1)
+  )
+  data.frame(
+    PackratFormat = NA,
+    PackratVersion = NA,
+    RVersion = renv$R$Version,
+    Repos = paste(repos, collapse = ",")
+  )
+}
+
+packratPackages <- function(packages) {
+  out <- lapply(packages, packratPackage)
+  names(out) <- NULL
+  out <- compact(out)
+  out <- lapply(out, as.data.frame, stringsAsFactors = FALSE)
+  rbind_fill(out)
+}
+
+packratPackage <- function(pkg) {
+  # Don't include renv itself
+  if (identical(pkg$Package, "renv")) {
+    return(NULL)
+  }
+
+  # Convert renv sources to packrat sources
+  # https://github.com/rstudio/renv/blob/0.17.2/R/snapshot.R
+  if (identical(pkg$Repository, "CRAN")) {
+    pkg$Source <- "CRAN"
+  } else if (pkg$Source == "unknown") {
+    pkg$Source <- "source"
+  }
+
+  # Remove Remote fields that renv adds for "standard" installs from CRAN
+  if (identical(pkg$RemoteType, "standard")) {
+    pkg <- pkg[!grepl("^Remote", names(pkg))]
+  }
+
+  # Drop hash since packrat hash != renv hash
+  pkg$Hash <- NULL
```

Comment:
One worry I have is that, if I understand correctly, rsconnect tries to recompute the hash from the package DESCRIPTION files, but in this scenario we can't assume those DESCRIPTION files are available (or match whatever is recorded in the renv lockfile).

This might warrant some further discussion.

Reply:
Hashes are recalculated by Connect + Packrat at restore-time using the `DESCRIPTION` files included in the `packrat/desc` directory.

When we have only a `manifest.json`, the entire `packrat` directory is reconstructed using the package descriptions contained in the manifest.

The most straightforward approach is for `rsconnect` to stop including the `packrat` directory in the bundle and have Connect rely only on the `manifest.json`. The manifest structure is unchanged (including the full package description). That makes `rsconnect` driven deploys more similar to Git-backed deploys and `rsconnect-python` driven deploys.

Reply:
What if we don't have any DESCRIPTION files available? For example, suppose that the package records declared in `renv.lock` aren't installed at all. More broadly, we might not be able to assume that `renv.lock` and the user's R library are actually in sync.

In theory, we could obtain those DESCRIPTION files based on the package records provided (I feel like we had some discussions about a "meta-CRAN" for something like this).

Reply:
I think we assume that you can only deploy content that you can run locally; i.e. all the `DESCRIPTION` files are available. But it does seem wrong to get these `DESCRIPTION` files from the current library, not the library that renv used.

Reply:
But we can at least take the `DESCRIPTION`s from renv library, not the system library.